internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+
+	var wg sync.WaitGroup
+
+	checker := func(fn CheckHandlerFunc) {
+		resp, err := fn(ctx)
+		resultChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}
+
+	wg.Add(1)
+	go func() {
+	outer:
+		for _, handler := range handlers {
+			fn := handler // capture loop var
+
+			select {
+			case limiter <- struct{}{}:
+				wg.Add(1)
+				go checker(fn)
+			case <-ctx.Done():
+				break outer
+			}
+		}
+
+		wg.Done()
+	}()
+
+	return func() {
+		wg.Wait()
+		close(limiter)
+	}
+}
+
+// union implements a CheckFuncReducer that requires any of the provided CheckHandlerFunc to resolve
+// to an allowed outcome. The first allowed outcome causes premature termination of the reducer.
+func union(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	cctx, cancel := context.WithCancel(ctx)
+	resultChan := make(chan checkOutcome, len(handlers))
+
+	drain := resolver(cctx, concurrencyLimit, resultChan, handlers...)
+
+	defer func() {
+		cancel()
+		drain()
+		close(resultChan)
+	}()
+
+	for i := 0; i < len(handlers); i++ {
+		select {
+		case result := <-resultChan:
+			if result.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, result.err
+			}
+
+			if result.resp.GetAllowed() {
+				return result.resp, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	return &openfgapb.CheckResponse{Allowed: false}, nil
+}
+
+// intersection implements a CheckFuncReducer that requires all of the provided CheckHandlerFunc to resolve
+// to an allowed outcome. The first falsey or erroneous outcome causes premature termination of the reducer.
+func intersection(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	cctx, cancel := context.WithCancel(ctx)
+	resultChan := make(chan checkOutcome, len(handlers))
+
+	drain := resolver(cctx, concurrencyLimit, resultChan, handlers...)
+
+	defer func() {
+		cancel()
+		drain()
+		close(resultChan)
+	}()
+
+	for i := 0; i < len(handlers); i++ {
+		select {
+		case result := <-resultChan:
+
+			if result.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, result.err
+			}
+
+			if !result.resp.GetAllowed() {
+				return result.resp, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	return &openfgapb.CheckResponse{Allowed: true}, nil
+}
+
+// exclusion implements a CheckFuncReducer that requires a 'base' CheckHandlerFunc to resolve to an allowed
+// outcome and a 'sub' CheckHandlerFunc to resolve to a falsey outcome. The base and sub computations are
+// handled concurrently relative to one another.
+func exclusion(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	if len(handlers) != 2 {
+		panic(fmt.Sprintf("expected two rewrite operands for exclusion operator, but got '%d'", len(handlers)))
+	}
+
+	limiter := make(chan struct{}, concurrencyLimit)
```

Comment:
I'm not sure I understand.. The limiter here does the same thing it does in `intersection` and `union`. It limits the total number of concurrent goroutines can be inflight for the evaluation of the `CheckHandlerFunc` passed in. For the `exclusion` case if `concurrencyLimit` is 1, then only the base or subtract handlers can be run at a time (the other has to wait for the other to finish).

For intersection and union, if the number of `CheckHandlerFunc` is greater than the `concurrencyLimit`, then we only evaluate up to `concurrencyLimit` handlers at a time, and the others have to wait until those are complete.

Reply:
> For the exclusion case if concurrencyLimit is 1, then only the base or subtract handlers can be run at a time (the other has to wait for the other to finish).

But why would I even set it to 1? Could that happen?

I don't think we need a limiter in the exclusion case because we know that we will always run 2 goroutines. If each branch of `base` and `sub` are more complex, those will be rate-limited separately.

On a separate note, if we do keep the limiter: do we need to close it in the `defer` function?

Reply:
The function accepts a `concurrencyLimit`. The function doesn't assume that variable is a certain range (e.g. > 1). So if you call this function with a configured concurrency limit of `1` (which may be possible if we allow concurrency configurations in the future), then this code works regardless of what the limit is. The limit could be 1 or it could be 1000, and this code still works.